### PR TITLE
WIP - parse files only when sourcemap is needed

### PR DESCRIPTION
**Pull request by roironn**
Still needs tests, and probably some refactoring - but I'd be happy if ya'll can have a look at the general idea until I'll add them :)

### CODE REVIEW

File: packages/workspace/src/parser/parse.ts [hunk 1]
```diff
@@ -53,7 +53,7 @@ export const parse = async (
   naclFile: Buffer,
   filename: string,
   functions: Functions = {},
-): Promise<ParseResult> => {
+): Promise<Required<ParseResult>> => {
```
Comment (hadard) [inline packages/workspace/src/parser/parse.ts:56]: what `Required` means here?

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 2]
```diff
@@ -97,10 +99,15 @@ const buildNaclFilesSource = (
...
     ...getStaticFilesFunctions(staticFileSource), // add future functions here
   }
-  const parseNaclFile = async (naclFile: NaclFile): Promise<ParseResult> => {
-    const key = { filename: naclFile.filename, lastModified: naclFile.timestamp || Date.now() }
+  const cacheResultKey = (filename: string, timestamp?: number): ParseResultKey => ({
+    filename,
+    lastModified: timestamp ?? Date.now(),
+  })
+
+  const parseNaclFile = async (naclFile: NaclFile, forceParse = false): Promise<ParseResult> => {
```
Comment (hadard) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:107]: I really don't like this pattern where we need to pass boolean of `forceParse` - smells wrong

Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts]: 👍

File: packages/workspace/src/parser/parse.ts (hunk 1 above)
Comment (roironn) [reply packages/workspace/src/parser/parse.ts:56]: It means that all of the optional Type fields are present. (In our case - when invoking the actual parse function the source map will always be returned. This prevents us from having to support 'never gonna be executed' code branches in getSourceMap

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 3]
```diff
@@ -97,28 +99,58 @@ const buildNaclFilesSource = (
...
-        Buffer.from(naclFile.buffer), naclFile.filename,
-        functions,
-      )
-      await cache.put(key, parseResult)
+  const cacheResultKey = (filename: string, timestamp?: number): ParseResultKey => ({
+    filename,
+    lastModified: timestamp ?? Date.now(),
+  })
+
+  const parseNaclFileBuffer = async (naclFile: NaclFile): Promise<Required<ParseResult>> => (
```
Comment (hadard) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:107]: super nit: I think the `(` and `)` around `parse` are redundant

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 4]
```diff
@@ -97,28 +99,58 @@ const buildNaclFilesSource = (
...
+    parse(Buffer.from(naclFile.buffer), naclFile.filename, functions)
+  )
+
+  const parsedNaclFileFromParseResult = (
```
Comment (hadard) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:111]: super super nit: rename to `toParsedNaclFile`?  the problem I have with the name is that we actually takes `NaclFile` and `ParseResult` as the input and not just `ParseResult`

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 5]
```diff
@@ -97,28 +99,58 @@ const buildNaclFilesSource = (
...
+    filename: string,
+    change: AdditionDiff<Element>,
+  ): Promise<ParsedNaclFile> => {
+    const elements = [(change as AdditionDiff<Element>).data.after]
+    const parsed = {
+      timestamp: Date.now(),
+      filename,
+      elements: _.keyBy(elements, e => e.elemID.getFullName()),
+      errors: [],
     }
```
Comment (hadard) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:141]: this is sort of duplication of `parsedNaclFileFromParseResult`, I think we should use it here

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 6]
```diff
@@ -97,28 +99,58 @@ const buildNaclFilesSource = (
...
-    return parseResult
+    const key = cacheResultKey(parsed.filename, parsed.timestamp)
+    await cache.put(key, { elements, errors: [] })
```
Comment (hadard) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:143]: this pattern of creating the `cacheResultKey` based on `ParsedNaclFile` and then using `cache.put` is repating few times, I think instead of having the function `cacheResultKey` we can declare `updateCache(ParsedNaclFile, ParseResult)`

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 7]
```diff
@@ -127,11 +159,10 @@ const buildNaclFilesSource = (
       _.isUndefined
     ) as NaclFile[]
 
-  const buildNaclFilesState = async (newNaclFiles: NaclFile[], current: ParsedNaclFileMap):
+  const buildNaclFilesState = async (newNaclFiles: ParsedNaclFile[], current?: ParsedNaclFileMap):
```
Comment (hadard) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:162]: not sure I understand why this change is needed, looks like the call to `parseNaclFiles` just moved to a different place, no?

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 8]
```diff
@@ -171,29 +204,31 @@ const buildNaclFilesSource = (
...
-      }
-      return (await parseNaclFile({ filename, buffer })).sourceMap
+    const key = cacheResultKey(parsedNaclFile.filename, parsedNaclFile.timestamp)
+    const cachedResult = await cache.get(key)
+    if (cachedResult && cachedResult.sourceMap) {
+      return cachedResult.sourceMap
     }
-
-    return cachedParsedResult.sourceMap
+    log.debug('expected to find source map for filename %s, going to re-parse', filename)
```
Comment (hadard) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:212]: I think the `log` print is not true anymore

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 9]
```diff
@@ -171,29 +204,31 @@ const buildNaclFilesSource = (
...
+    const naclFile = (await naclFilesStore.get(filename))
+    if (_.isUndefined(naclFile)) {
+      log.error('failed to find %s in NaCl file store', filename)
+      return new SourceMap()
+    }
+    const parsedResult = await parseNaclFileBuffer(naclFile)
+    await cache.put(key, parsedResult)
+    return parsedResult.sourceMap
```
Comment (hadard) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:220]: IMO better to call here `parseNaclFile` than `put` explicit in the `cache`
I think this will also mean `parseNaclFileBuffer` can be deleted and we just have `parseNaclFile` that being used. No?

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 10]
```diff
@@ -171,29 +204,31 @@ const buildNaclFilesSource = (
...
+    ...naclFiles: NaclFile[]
+  ): Promise<void> => {
     const [emptyNaclFiles, nonEmptyNaclFiles] = _.partition(
       naclFiles,
       naclFile => _.isEmpty(naclFile.buffer.trim())
     )
     await Promise.all(nonEmptyNaclFiles.map(naclFile => naclFilesStore.set(naclFile)))
     await Promise.all(emptyNaclFiles.map(naclFile => naclFilesStore.delete(naclFile.filename)))
-    // Swap state
-    state = buildNaclFilesState(naclFiles, (await getState()).parsedNaclFiles)
```
Comment (hadard) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:196]: seems like this was removed to where we call `setNaclFiles` what is the use case where we don't build state? what is the motivation of moving this out of `setNaclFiles`?

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts (hunk 3 above)
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:107]: IMO it really helps readability wise - but sure, why not :)

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts (hunk 4 above)
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:111]: 👍

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts (hunk 5 above)
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:141]: Its the same output, but if you consider the input we don't have a naclFile nor a parseResult here, which would require us to 'fake' them in the input (and we can't even fake the nacl file since we don't have a buffer)

I can change the `toParsedNaclFile` method params to be filename and an optional timestamp, but that would make it a bit cumbersome...

Thoughts?

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts (hunk 7 above)
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:162]: Having the original param here set as NaclFile forced use to in to create the state elements by parsing the files buffer. Without this - we won't be able to avoid parsing the 'new' updates...

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts (hunk 8 above)
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:212]: 👍

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts (hunk 9 above)
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:220]: 👍 
Please note that in order to achieve this I changed the return type of parseNaclFile to ParseResult (since the 
 `ParsedNaclFile` interface doesn't expose the sourceMap in order to evacuate it from memory) and wrap some calls in toParsedNaclFile.

I actually think it worked out rather nicely, let me know what you think.

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts (hunk 6 above)
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:143]: After the changes made in the PR we only call cache.put in two places, and in one of them we don't have a parsed nacl file (only parse result) - so I think considering the new chanceResultKey function, the updateCache method suggested it not really needed - thoughts?

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts (hunk 10 above)
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:196]: The reason for this is the change made to the `buildNaclFilesState`. (Explained in above comment)
Since setNaclFile receives naclFiles (which does not expose the buffer interface to allow the buffers to be evacuated from memory)  we can't invoke buildNaclFilesState from that method.

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 11]
```diff
@@ -227,9 +255,17 @@ const buildNaclFilesSource = (
...
-            const buffer = await updateNaclFileData(await getNaclFileData(filename),
-              fileChanges, functions)
-            return { filename, buffer }
+            const naclFileData = await getNaclFileData(filename)
+            const buffer = await updateNaclFileData(naclFileData, fileChanges, functions)
+            const shouldNotParse = _.isEmpty(naclFileData)
+              && fileChanges.length === 1
+              && fileChanges[0].action === 'add'
+              && isElement(fileChanges[0].data.after)
+            const naclFile = { filename, buffer }
```
Comment (hadard) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:264]: nit: `naclFile` is only used a 3 lines after and once IMO more readable to have  `await parseNaclFile({ filename, buffer })`
